File: Desktop/irctc-concurrency/fastapi-trains/app/waitlist/repository.py
from sqlalchemy.orm import Session
from sqlalchemy import text


class WaitlistRepository:

    def __init__(self, db: Session):
        self.db = db

    def get_waitlist_count(self, train_id: int, journey_date: str):
        return self.db.execute(
            text("""
                SELECT COUNT(*) as waiting_count
                FROM waitlist
                WHERE train_id = :train_id
                AND journey_date = :journey_date
                AND status = 'WAITING'
            """),
            {"train_id": train_id, "journey_date": journey_date}
        ).fetchone()

    def find_existing(self, user_id: int, train_id: int, journey_date: str):
        return self.db.execute(
            text("""
                SELECT id, position, status FROM waitlist
                WHERE user_id = :user_id
                AND train_id = :train_id
                AND journey_date = :journey_date
                AND status = 'WAITING'
                ORDER BY created_at DESC
                LIMIT 1
            """),
            {"user_id": user_id, "train_id": train_id, "journey_date": journey_date}
        ).fetchone()

    def insert_waitlist(self, user_id: int, train_id: int, journey_date: str):
        self.db.execute(
            text("""
                INSERT INTO waitlist (user_id, train_id, journey_date, position, status)
                VALUES (:user_id, :train_id, :journey_date,
                    (SELECT COALESCE(MAX(position), 0) + 1
                    FROM waitlist
                    WHERE train_id = :train_id
                    AND journey_date = :journey_date
                    AND status = 'WAITING'),'WAITING')
            """),
            {"user_id": user_id, "train_id": train_id, "journey_date": journey_date}
        )
        self.db.commit()

    def get_user_waitlist(self, user_id: int):
        return self.db.execute(
            text("""
                SELECT w.id, w.train_id, w.journey_date,
                    w.position, w.status, w.created_at,
                    t.train_name, t.source, t.destination,
                    u.name as user_name
                FROM waitlist w
                JOIN trains t ON t.id = w.train_id
                JOIN users u ON u.id = w.user_id
                WHERE w.user_id = :user_id
                AND w.status = 'WAITING'
                ORDER BY w.position ASC
            """),
            {"user_id": user_id}
        ).fetchall()

    def confirm_next_atomically(self, train_id: int, journey_date: str):
        result = self.db.execute(
            text("""
                WITH locked_waitlist AS (
                    SELECT id, user_id, position
                    FROM waitlist
                    WHERE train_id  = :train_id
                      AND journey_date = :journey_date
                      AND status = 'WAITING'
                    ORDER BY position ASC
                    LIMIT 1
                    FOR UPDATE SKIP LOCKED
                ),
                locked_seat AS (
                    SELECT s.id AS seat_id, s.seat_number
                    FROM seats s
                    WHERE s.train_id = :train_id
                      AND NOT EXISTS (
                          SELECT 1 FROM bookings b
                          WHERE b.seat_id    = s.id
                            AND b.journey_date = :journey_date
                            AND b.status = 'CONFIRMED'
                      )
                    LIMIT 1
                    FOR UPDATE SKIP LOCKED
                ),
                new_booking AS (
                    INSERT INTO bookings (user_id, train_id, seat_id, journey_date, status, booked_at)
                    SELECT lw.user_id, :train_id, ls.seat_id, :journey_date, 'CONFIRMED', NOW()
                    FROM locked_waitlist lw, locked_seat ls
                    RETURNING id AS booking_id, seat_id, user_id
                ),
                updated_waitlist AS (
                    UPDATE waitlist
                    SET status = 'CONFIRMED'
                    FROM locked_waitlist lw
                    WHERE waitlist.id = lw.id
                    RETURNING waitlist.id AS waitlist_id
                )
                SELECT
                    nb.booking_id,
                    nb.user_id,
                    nb.seat_id,
                    ls.seat_number,
                    uw.waitlist_id
                FROM new_booking nb
                JOIN locked_seat   ls ON ls.seat_id = nb.seat_id
                JOIN updated_waitlist uw ON TRUE
            """),
            {"train_id": train_id, "journey_date": journey_date}
        ).fetchone()

        self.db.commit()
        return result

    def cleanup_user_waitlist(self, user_id: int, train_id: int, journey_date: str):
        self.db.execute(
            text("""
                DELETE FROM waitlist
                WHERE user_id = :user_id
                AND train_id = :train_id
                AND journey_date = :journey_date
                AND status = 'CONFIRMED'
            """),
            {"user_id": user_id, "train_id": train_id, "journey_date": journey_date}
        )
        self.db.commit()

File: Desktop/irctc-concurrency/gin-booking/internal/booking/repository.go
package booking

import (
	"database/sql"
	"time"
)

type Repository struct {
	db *sql.DB
}

func NewRepository(db *sql.DB) *Repository {
	return &Repository{db: db}
}

func (r *Repository) GetSeatByID(seatID int) (*Seat, error) {
	seat := &Seat{}
	err := r.db.QueryRow(`
		SELECT id, train_id, seat_number, seat_type, is_available
		FROM seats WHERE id = $1`, seatID).Scan(&seat.ID, &seat.TrainID, &seat.SeatNumber, &seat.SeatType, &seat.IsAvailable)
	if err != nil {
		return nil, err
	}
	return seat, nil
}

func (r *Repository) LockSeat(seatID int, tx *sql.Tx) error {
	_, err := tx.Exec(`UPDATE seats SET is_available = false WHERE id = $1`, seatID)
	return err
}

func (r *Repository) DecrementAvailableSeats(trainID int, tx *sql.Tx) error {
	_, err := tx.Exec(`UPDATE trains
		SET available_seats = available_seats - 1
		WHERE id = $1 AND available_seats > 0`, trainID)
	return err
}

func (r *Repository) CreateBooking(userID, trainID, seatID int, journeyDate, status string, tx *sql.Tx) (int, error) {
	var bookingID int
	err := tx.QueryRow(`INSERT INTO bookings (user_id, train_id, seat_id, journey_date, status, booked_at)
		VALUES ($1, $2, $3, $4, $5, $6)
		RETURNING id`, userID, trainID, seatID, journeyDate, status, time.Now()).Scan(&bookingID)
	return bookingID, err
}

func (r *Repository) GetBookingByID(bookingID int) (*Booking, error) {
	booking := &Booking{}
	err := r.db.QueryRow(`
		SELECT id, user_id, train_id, seat_id, journey_date, status, booked_at
		FROM bookings WHERE id = $1`, bookingID).Scan(&booking.ID, &booking.UserID, &booking.TrainID, &booking.SeatID, &booking.JourneyDate,
		&booking.Status, &booking.BookedAt,
	)
	if err != nil {
		return nil, err
	}
	return booking, nil
}

func (r *Repository) CancelBooking(bookingID int) error {
	_, err := r.db.Exec(`
		UPDATE bookings SET status = 'CANCELLED' WHERE id = $1`, bookingID)
	return err
}

func (r *Repository) UnlockSeat(seatID int) error {
	_, err := r.db.Exec(`
		UPDATE seats SET is_available = true WHERE id = $1`, seatID)
	return err
}

func (r *Repository) UnlockSeatTx(seatID int, tx *sql.Tx) error {
	_, err := tx.Exec(`UPDATE seats SET is_available = true WHERE id = $1`, seatID)
	return err
}

func (r *Repository) DeleteBooking(bookingID int, tx *sql.Tx) error {
	_, err := tx.Exec(`DELETE FROM bookings WHERE id = $1`, bookingID)
	return err
}

func (r *Repository) IncrementAvailableSeats(trainID int, tx *sql.Tx) error {
	_, err := tx.Exec(`
		UPDATE trains
		SET available_seats = available_seats + 1
		WHERE id = $1`, trainID)
	return err
}

func (r *Repository) GetBookingsByUser(userID int) ([]Booking, error) {
	rows, err := r.db.Query(`
		SELECT b.id, b.user_id, b.train_id, b.seat_id, b.journey_date, b.status, b.booked_at,s.seat_number
		FROM bookings b
		JOIN seats s ON s.id = b.seat_id
		WHERE b.user_id = $1
		ORDER BY b.booked_at DESC`, userID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var bookings []Booking
	for rows.Next() {
		var b Booking
		err := rows.Scan(&b.ID, &b.UserID, &b.TrainID, &b.SeatID, &b.JourneyDate, &b.Status, &b.BookedAt, &b.SeatNumber)
		if err != nil {
			continue
		}
		bookings = append(bookings, b)
	}
	return bookings, nil
}

func (r *Repository) IsSeatAvailableForDate(seatID int, journeyDate string) (bool, error) {
	var count int
	err := r.db.QueryRow(`
		SELECT COUNT(*) FROM bookings
		WHERE seat_id = $1
		AND journey_date = $2
		AND status = 'CONFIRMED'
	`, seatID, journeyDate).Scan(&count)
	if err != nil {
		return false, err
	}
	return count == 0, nil
}
func (r *Repository) GetDepartureTime(trainID int) (string, error) {
	var departureTime string
	err := r.db.QueryRow(`
        SELECT departure_time FROM trains WHERE id = $1`, trainID).Scan(&departureTime)
	return departureTime, err
}
func (r *Repository) GetAvailableSeats(trainID int) (int, error) {
	var count int
	err := r.db.QueryRow(`SELECT available_seats FROM trains WHERE id = $1`, trainID).Scan(&count)
	return count, err
}
